copy/copy.go

```diff
@@ -487,41 +490,84 @@ func (c *copier) copyMultipleImages(ctx context.Context, policyContext *signatur
 	}
-	c.Printf("Copying %d of %d images in list\n", imagesToCopy, len(instanceDigests))
-	updates := make([]manifest.ListUpdate, len(instanceDigests))
+	c.Printf("Copying %d of %d images in list\n", imagesToCopy, imagesToCopy)
+	updates := make([]manifest.ListUpdate, imagesToCopy)
 	instancesCopied := 0
-	for i, instanceDigest := range instanceDigests {
-		if options.ImageListSelection == CopySpecificImages &&
-			!slices.Contains(options.Instances, instanceDigest) {
-			update, err := updatedList.Instance(instanceDigest)
+	processDigestsTimes := 1
+	if options.ReplicateZstd {
+		processDigestsTimes = 2
+	}
+	for iter := 0; iter < processDigestsTimes; iter++ {
+		for i, instanceDigest := range instanceDigests {
+			if iter == 1 && options.ReplicateZstd {
+				i = i + 2
+				// Change compression to zstd
+				options.DestinationCtx.CompressionFormat = &compression.Zstd
+				level := 20
+				options.DestinationCtx.CompressionLevel = &level
+				c.compressionFormat = &compression.Zstd
+				c.compressionLevel = &level
+			}
+			if options.ImageListSelection == CopySpecificImages &&
+				!slices.Contains(options.Instances, instanceDigest) {
+				update, err := updatedList.Instance(instanceDigest)
+				if err != nil {
+					return nil, err
+				}
+				logrus.Debugf("Skipping instance %s (%d/%d)", instanceDigest, i+1, len(instanceDigests))
+				// Record the digest/size/type of the manifest that we didn't copy.
+				updates[i] = update
+				continue
+			}
+			logrus.Debugf("Copying instance %s (%d/%d)", instanceDigest, i+1, len(instanceDigests))
+			c.Printf("Copying image %s (%d/%d)\n", instanceDigest, instancesCopied+1, imagesToCopy)
+			unparsedInstance := image.UnparsedInstance(c.rawSource, &instanceDigest)
+			updatedManifest, updatedManifestType, updatedManifestDigest, parsedImage, err := c.copyOneImage(ctx, policyContext, options, unparsedToplevel, unparsedInstance, &instanceDigest)
 			if err != nil {
-				return nil, err
+				return nil, fmt.Errorf("copying image %d/%d from manifest list: %w", instancesCopied+1, imagesToCopy, err)
+			}
+			instancesCopied++
+			// Record the result of a possible conversion here.
+			update := manifest.ListUpdate{
+				Digest:    updatedManifestDigest,
+				Size:      int64(len(updatedManifest)),
+				MediaType: updatedManifestType,
+			}
+			if iter == 1 && options.ReplicateZstd {
+				srcImage, err := parsedImage.OCIConfig(ctx)
+				if err != nil {
+					return nil, err
+				}
+				blobInfo := parsedImage.ConfigInfo()
+				updatedPlatform := imgspecv1.Platform{Architecture: srcImage.Architecture, OS: srcImage.OS, OSVersion: srcImage.OSVersion, OSFeatures: srcImage.OSFeatures, Variant: srcImage.Variant}
+				instancesCopied++
+				annotations := make(map[string]string)
+				for k, v := range blobInfo.Annotations {
+					annotations[k] = v
+				}
+				// set the standard zstd annotations if annotation for zstd replication was set
+				annotations[internalManifest.OCI1InstanceAnnotationCompressionZSTD] = "true"
```

Comment:
The specific annotation name (and even that this is recorded in an _annotation_, and not some other field) is OCI-specific and should be implemented in the OCI-specific code.

---

Actually, we should add that annotation also if the user explicitly requests conversion via `DestinationCtx.CompressionFormat`. That’s a subset of this feature that is somewhat higher-priority because the API already exists. Adding that to the PR list.